src/Type/Doctrine/Query/QueryType.php

```diff
@@ -3,19 +3,21 @@
 namespace PHPStan\Type\Doctrine\Query;
 
 use PHPStan\TrinaryLogic;
-use PHPStan\Type\ObjectType;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\MixedType;
 use PHPStan\Type\Type;
 
 /** @api */
-class QueryType extends ObjectType
+class QueryType extends GenericObjectType
 {
 
 	/** @var string */
 	private $dql;
 
-	public function __construct(string $dql)
+	public function __construct(string $dql, ?Type $resultType = null)
 	{
-		parent::__construct('Doctrine\ORM\Query');
+		$resultType = $resultType ?? new MixedType();
+		parent::__construct('Doctrine\ORM\Query', [$resultType]);
```

Comment:
Taking the `$dql` as a string here forces us to create a separate QueryType instance for each alternative query. This becomes a problem with the generic parameter, because instead of typing `$em->createQuery('select e from ' . (rand(0,1) ? 'Foo' : 'Bar'))` as `Query<Foo|Bar>`, we are forced to type it as `Query<Foo>|Query<Bar>`. The former is more useful in some cases: https://phpstan.org/r/7ce3e2f6-f544-4637-87bb-57a96e7e8ba2

Changing `string $dql` to `Type|string $dql` would work, but it would break the compatibility of the getDql() method. Alternatively, we could introduce a new type: `GenericQueryType`, and use this type instead of `QueryType`. Would this be acceptable with the backward compatibility promise ?

Reply:
(not very critical though, since createQuery() rarely receives a union of constant strings in practice)

Reply:
Yeah, I think we can disregard these cases for now to make it simple :)